### PR TITLE
T9999: Add wgIncidentReportingInactiveServices

### PR DESCRIPTION
**Pull request by redbluegreenhat**
https://phabricator.miraheze.org/T9999

### CODE REVIEW

**Comment by Universal-Omega** [comment]
version needs bumped and entry in changelog

**Comment by redbluegreenhat** [comment]
done.

File: extension.json [hunk 1]
```diff
@@ -5,7 +5,7 @@
 		"Universal Omega"
 	],
 	"url": "https://github.com/miraheze/IncidentReporting",
-	"version": "1.1.9",
+	"version": "1.1.10",
```
Comment (Universal-Omega) [inline extension.json:8]: Should be 1.2.0 I think.

**Comment by Reception123** [comment]
I may be doing something wrong (maybe the old changes didn't get disapplied on test131) but this still seems to erase the name? (though as I say this could be the result of a mistake with testing)

**Comment by Reception123** [comment]
In my testing it looks like this https://phabricator.miraheze.org/T9999#209676 though as I said it might be the older version that didn't disapply properly

**Comment by Universal-Omega** [comment]
No, this PR seems wrong. It should be updating SpecialIncidentReports. I don't think the file touched here needs to be.

**Comment by Universal-Omega** [comment]
Also you don't need to set strict for in_array here.

**Comment by Reception123** [comment]
@Universal-Omega that's what was tried with https://github.com/miraheze/IncidentReporting/pull/55 but it made the inactive service incidents disappear/threw an error when trying to access them

**Comment by Universal-Omega** [comment]
> @Universal-Omega that's what was tried with #55 but it made the inactive service incidents disappear/threw an error when trying to access them

That was correct, for the most part, except that https://github.com/miraheze/IncidentReporting/blob/b0ea18035c69694bc65d8bd2c310c78163614506/includes/SpecialIncidentReports.php#L176 still needs those added, only the dropdown doesn't.

**Comment by redbluegreenhat** [comment]
If this doesn't work then I'm officially out of ideas on how to implement this, short of rewriting this extension from scratch.

@Universal-Omega SpecialIncidentReports.php didn't seem to have anything relevant. When editing (and creating?) tasks, it creates a new instance of the form factory (https://github.com/miraheze/IncidentReporting/blob/master/includes/SpecialIncidentReports.php#L69), which seems to use the irServices variable in a dropdown(?) (https://github.com/miraheze/IncidentReporting/blob/master/includes/IncidentReportingFormFactory.php#L115) when selecting the service related to an incident report.

The irServices var is an array containing the formatted name of the service ("-" character removed, basically, https://github.com/miraheze/IncidentReporting/blob/master/includes/IncidentReportingFormFactory.php#L56). The idea of this PR is preventing the service from being added to the irServices variable, thus preventing it from being an option in the dropdown when selecting services in an incident report.

**Comment by Universal-Omega** [comment]
Maybe it needs done in a couple different places...

This extension is a mess in this part. Partly my fault in my earlier days before my code standards improved more. Alot of mess was done with #17/#18/#22 (adding statistics) which I'll cleanup later also.

**Comment by Universal-Omega** [comment]
Wait, I know how to fix this. Use array_diff for options, rather than directly excluding from $irServices. I think that will need done in both places though.

**Comment by redbluegreenhat** [comment]
> Wait, I know how to fix this. Use array_diff for options, rather than directly excluding from $irServices.

Will add that later, though it just seems like another way of doing what this PR is already doing?

I just had a revelation. @Reception123 Nothing in this PR should affect the landing page, does the broken task also appear in the landing page (https://meta.miraheze.org/wiki/Special:IncidentReports, which it shouldn't as nothing related to that is touched in this PR), or does this only affect incident reports of  services marked as inactive in the settings (like https://meta.miraheze.org/wiki/Special:IncidentReports/1). I could see it freaking out in the last case.

The extension checks if any extra parameters in the form of URL endpoints are passed (https://github.com/miraheze/IncidentReporting/blob/b45bfe4d5949cb970771e65f9e7699cee4573793/includes/SpecialIncidentReports.php#L24, At least that's how I tried to explain it, more commonly know as subpages in MediaWiki speak). It checks if anything is passed (https://github.com/miraheze/IncidentReporting/blob/b45bfe4d5949cb970771e65f9e7699cee4573793/includes/SpecialIncidentReports.php#L40).

If not, it shows the landing page, which should be unaffected.

If it is, it runs a few more checks. If the user has the (editincidents) permission (https://github.com/miraheze/IncidentReporting/blob/b45bfe4d5949cb970771e65f9e7699cee4573793/includes/SpecialIncidentReports.php#L43). The important thing here, is that, in both cases, the page is generated using the form factory, just that, if you have permissions, it allows editing the report.

So that's how incident reports are broken with this change. Given that, I have a few ideas (assuming I'm right and that this is what's going on):

Check if the user has the editincidents permission before removing services from the irServices variable: This would hide those services when editing/creating reports, allow users without the permission to read those reports, and break those reports for users with that permission. Probably the easiest "fix", but not really acceptable for a project like this, I suppose.

I don't really know how to fix this without breaking reports of inactive services for at-least users with permissions.

**Comment by redbluegreenhat** [comment]
I should also note that the reason the extension freaks out is: https://github.com/miraheze/IncidentReporting/blob/afeb26e8f9fd92069849c4e6de29201eb5323430/includes/IncidentReportingFormFactory.php#L204

**Comment by redbluegreenhat** [comment]
@Universal-Omega I just realized what you meant with array_diff, yeah that's the way to fix this, forget the above.

File: extension.json (hunk 1 above)
Comment (redbluegreenhat) [reply extension.json:8]: Done.

**Comment by Reception123** [comment]
Same disclaimer (regarding potentially wrong testing) as before but I still see it missing

**Comment by redbluegreenhat** [comment]
> Same disclaimer (regarding potentially wrong testing) as before but I still see it missing

Is test131 betaheze? I guess users with editincidents always see the edit form, so the array_diff solution doesn't fix this for them, reports of inactive services are broken at least for them.

**Comment by Reception123** [comment]
Yes, test131 is betaheze.

**Comment by redbluegreenhat** [comment]
I have another idea on how to implement this: CSS. We could use CSS on that page to hide the select elements of inactive services in the form. We could attach to the PageBeforeDisplay hook, check if the page is Special:IncidentReports, and add dynamically-generated CSS to hide those options.

**Comment by redbluegreenhat** [comment]
I'm going to open a new PR for the CSS idea.

**Pull request closed without merging**